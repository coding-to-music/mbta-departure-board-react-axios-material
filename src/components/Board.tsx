import React, { useEffect, useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import { IPrediction } from '../dataSource/IPredition';
import BoardEntry from './BoardEntry';

function Board(props: { data: IPrediction[] }) {
  const [departures, setDepartures] = useState([] as IPrediction[]);
  const { data } = props;

  useEffect(() => {
    const departuresFilter = data.filter(
      (pred) => pred.attributes.departure_time !== null,
    );

    const sortedDepartures = departuresFilter.sort(
      (a: IPrediction, b: IPrediction) => (
        new Date(a.attributes.departure_time).getTime()
        - new Date(b.attributes.departure_time).getTime()
      ),
    );

    setDepartures(sortedDepartures);
  }, [data]);

  const useStyles = makeStyles({
    table: {
      minWidth: 650,
    },
  });

  const classes = useStyles();

  return (
    <TableContainer component={Paper}>
      <Table className={classes.table} aria-label="Departure/Arrivals Board">
        <TableHead>
          <TableRow>
            <TableCell>
              <b>Carrier</b>
            </TableCell>
            <TableCell align="right">
              <b>Time</b>
            </TableCell>
            <TableCell align="right">
              <b>Destination</b>
            </TableCell>
            <TableCell align="right">
              <b>Train #</b>
            </TableCell>
            <TableCell align="right">
              <b>Track #</b>
            </TableCell>
            <TableCell align="right">
              <b>Status</b>
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {departures.map((departure: IPrediction) => (
            <BoardEntry key={departure.id} data={departure} />
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
}

export default Board;
