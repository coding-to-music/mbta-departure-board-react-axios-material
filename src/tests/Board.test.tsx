import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import axios from 'axios';
import pretty from 'pretty';
import Board from '../components/Board';
import fakePredictions, { trips } from './fakePredictions';

beforeEach(() => {
  jest.clearAllMocks();
});

jest.mock('axios');

test('Board renders correctly', async () => {
  const el = document.createElement('div');
  (axios.get as jest.Mock).mockImplementation(() => Promise.resolve({ data: { ...trips } }));
  await act(async () => {
    ReactDOM.render(<Board data={fakePredictions} />, el);
  });

  expect(pretty(el.innerHTML)).toMatchSnapshot();
});

test('Board display props', async () => {
  const prediction = fakePredictions[0];
  // eslint-disable-next-line max-len
  (axios.get as jest.Mock).mockImplementation(() => Promise.resolve({ data: { ...trips } }));

  const el = document.createElement('div');
  await act(async () => {
    ReactDOM.render(<Board data={fakePredictions} />, el);
  });
  expect(el.innerHTML).toContain(prediction.attributes.status!);
});

test('Entry shpow up in departure section', async () => {
  // eslint-disable-next-line max-len
  (axios.get as jest.Mock).mockImplementation(() => Promise.resolve({ data: { ...trips } }));

  const el = document.createElement('div');
  await act(async () => {
    ReactDOM.render(<Board data={fakePredictions} />, el);
  });

  const departuresElements = el.getElementsByClassName('MuiTableRow-root departure');
  const entryDeparture = fakePredictions.filter((pred) => pred.attributes.stop_sequence === 1)[0];
  expect(departuresElements[0].innerHTML).toContain(entryDeparture.attributes.status);
});

test('Entry shpow up in arrival section', async () => {
  // eslint-disable-next-line max-len
  (axios.get as jest.Mock).mockImplementation(() => Promise.resolve({ data: { ...trips } }));

  const el = document.createElement('div');
  await act(async () => {
    ReactDOM.render(<Board data={fakePredictions} />, el);
  });

  const departuresElements = el.getElementsByClassName('MuiTableRow-root arrival');
  const entryArrival = fakePredictions.filter((pred) => pred.attributes.stop_sequence > 1)[0];
  expect(departuresElements[0].innerHTML).toContain(entryArrival.attributes.status);
});
