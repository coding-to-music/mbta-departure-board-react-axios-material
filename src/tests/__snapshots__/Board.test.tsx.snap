// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board renders correctly 1`] = `
"<div class=\\"MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded\\">
  <table class=\\"MuiTable-root makeStyles-table-6\\" aria-label=\\"Departure/Arrivals Board\\">
    <thead class=\\"MuiTableHead-root\\">
      <tr class=\\"MuiTableRow-root MuiTableRow-head\\">
        <th class=\\"MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-2\\" scope=\\"col\\"><b>Carrier</b></th>
        <th class=\\"MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-2 MuiTableCell-alignRight\\" scope=\\"col\\"><b>Time</b></th>
        <th class=\\"MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-2 MuiTableCell-alignRight\\" scope=\\"col\\"><b>Destination</b></th>
        <th class=\\"MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-2 MuiTableCell-alignRight\\" scope=\\"col\\"><b>Train #</b></th>
        <th class=\\"MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-2 MuiTableCell-alignRight\\" scope=\\"col\\"><b>Track #</b></th>
        <th class=\\"MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-2 MuiTableCell-alignRight\\" scope=\\"col\\"><b>Status</b></th>
      </tr>
    </thead>
    <tbody class=\\"MuiTableBody-root\\">
      <tr class=\\"MuiTableRow-root departure\\">
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-5 MuiTableCell-alignLeft\\">MTBA</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-5 MuiTableCell-alignRight\\">01:53 AM</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-5 MuiTableCell-alignRight\\">South Station</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-5 MuiTableCell-alignRight\\">G-10042</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-5 MuiTableCell-alignRight\\">TBD</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-5 MuiTableCell-alignRight\\">DEPARTURED</td>
      </tr>
      <tr class=\\"MuiTableRow-root\\">
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-7\\"></td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-7\\"></td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-7\\">Trains on way to South Station (1)</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-7\\"></td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-7\\"></td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-7\\"></td>
      </tr>
      <tr class=\\"MuiTableRow-root arrival\\">
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-5 MuiTableCell-alignLeft\\">MTBA</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-5 MuiTableCell-alignRight\\">01:53 AM</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-5 MuiTableCell-alignRight\\">South Station</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-5 MuiTableCell-alignRight\\">G-10042</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-5 MuiTableCell-alignRight\\">TBD</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-5 MuiTableCell-alignRight\\">ALL ABOARD</td>
      </tr>
    </tbody>
  </table>
</div>"
`;
